[quant] Add param calcs and tests for APoT observers

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79499
* #79530
* #79198
* __->__ #78905

### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

asl3 committed:
Add param calcs for APoT observers

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs for APoT observers"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -4,44 +4,135 @@
 class NonUniformQuantizationObserverBase(ObserverBase):
-    quant_min = None
-    quant_max = None
+    min_val: torch.tensor([])
```
i think this is just torch.Tensor

dzdang replied:
torch.Tensor will be deprecated in the future. torch.tensor's constructor is a bit more aligned with numpy array than torch.Tensor. i think this lint error is talking about something else

HDCharles commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -4,44 +4,135 @@
+        self.level_indices = torch.tensor([])
+        self.b = b
+        self.k = k
+        self.n = 0
```
generally a bit safer to initialize to None rather than a valid number if you want a placeholder value, both for k and n

HDCharles commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -4,44 +4,135 @@
+            self.n = self.b // self.k
+
+        if self.n == 0:
+            return (0.0, torch.tensor([]), torch.tensor([]))
```
its weird to me that when k isn't specified, we return an empty tensor, seems like k should just be a mandatory argument

dzdang replied:
@asl3 I feel we should throw an exception if `k == 0`

HDCharles commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -4,44 +4,135 @@
+        return (self.gamma, quantization_levels, level_indices)
+
+class APoTObserver(NonUniformQuantizationObserverBase):
     def __init__(
```
needs a description of the API vars meanings

HDCharles replied:
alternatively more descriptive variable names

dzdang replied:
@asl3  please link the paper and describe what the variables mean. I only know what gamma means because I read the paper

dzdang replied:
oh never mind, I see you did that above

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

HDCharles commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -6,18 +6,106 @@
+
+        quantization_levels3 = obs3_result[1]
+
+        quantization_levels3_test = torch.tensor([0.0000, 0.0022, 0.0045, 0.0067, 0.0089, 0.0112, 0.0134, 0.0156, 0.0179,
```
checking the values like this seems problematic. i.e. generally unit tests should verify that something is working through comparison to either a trivial answer or a reference implementation, i.e. its relatively straightforward to see that 2^0+2^-1=1.5 so gamma =2/3 makes sense and if something changes there, its a clear error. Here is a test with a reference comparison: https://www.internalfb.com/code/fbsource/[43fceb50fa4a87c2eded04c8624145a904c1a40c]/fbcode/caffe2/test/quantization/core/test_quantized_op.py?lines=78

on the other hand, i don't know where these hardcoded values came from, they're not trivial to generate and if they were generated by running the code itself, it defeats the purpose to compare with them. 

i'd either 
A) make a reference implementation to compare with 
B) look at higher level things like 'are all the values from 0 to 2**B-1 included in level_indices and does it have any repeats...etc
C) compare with itself when testing certain behaviors, i.e. signed=true vs signed=false should be the same quantization levels, just with both the positive and negative

HDCharles reviewed (changes requested):
high level looks good, some comments pertaining to code/test standards

dzdang commented:
it would be a good idea to also link the paper in the PR summary and reference the equation numbers in the paper that you implemented and the relevant sections. if you want the PR to be more self contained, you can also type out the equations in the PR summary

dzdang commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -4,27 +4,134 @@
+
+        # compute n and store as member variable
+        if self.k:
+            self.n = self.b // self.k
```
your implementation should guarantee n is an integer by ensuring b is divisible by k. APoT doesn't make sense if b % k != 0

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

HDCharles reviewed (commented):
looks good

nit: unsure if the commented out section is intended to be there

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,116 @@
-        obs = APoTObserver(t, t, t, 0, 0)
+        """
+        Test case 1
+        Test that error is thrown when k == 0
```
should `k` be a required argumnet?

jerryzh168 replied:
and should max_val and b be required argument as well?

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,116 @@
+        """
+        obs3 = APoTObserver(max_val=1.0, b=6, k=2)
+
+        obs3_result = obs3.calculate_qparams(signed=False)
```
wondering if it would be better to return a dict instead of a list here?

asl3 replied:
I think returning a tuple is best since we want to return gamma, quantization levels, and level indices which have different types (gamma is a float, the others are tensors).

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,116 @@
-        t = torch.Tensor()
-        obs = APoTObserver(t, t, t, 0, 0)
+        """
+        Test case 2
```
should we split them to different tests?

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,116 @@
+        levelindices_size_test2 = int(len(obs2_result[2]))
+        self.assertEqual(levelindices_size_test2, 16)
+
+        unique_elts = True
+        level_indices2_test_list = obs2_result[2].tolist()
+        for i in range(16):
+            if level_indices2_test_list.count(i) != 1:
+                unique_elts = False
+
+        self.assertTrue(unique_elts)
```
nit: I think the check can probably be a bit simpler, like `len(set(level_indices2_test_list)) == len(level_indices2_test_list)`?

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,116 @@
+        self.assertEqual(obs2_result[0], (2 / 3))
+
+        quantlevels_size_test2 = int(len(obs2_result[1]))
+        self.assertEqual(quantlevels_size_test2, 16)
```
nit: can we avoid hardcoding 16 here?

jerryzh168 commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,116 @@
 
-        with self.assertRaises(NotImplementedError):
-            obs._calculate_qparams()
+        self.assertEqual(obs2_result[0], (2 / 3))
```
similar for this one, can we avoid hardcoding 2 / 3?

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

dzdang commented:
btw, when you're fixing your PRs, you can convert it to a "draft" so it doesn't notify the reviewers. otherwise, we get email notifications for each PR that is updated

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_nonuniform_observer.py:
```diff
@@ -2,22 +2,137 @@
-        t = torch.Tensor()
-        obs = APoTObserver(t, t, t, 0, 0)
+    """
+        Test case 1
```
test

asl3 commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -35,29 +36,30 @@ def __init__(
-        if self.k:
-            self.n = self.b // self.k
+        # check for valid inputs of b, k
+        assert(self.k and self.k != 0)
```
yep k could be None

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

dzdang commented on torch/ao/quantization/experimental/observer.py:
```diff
@@ -4,30 +4,112 @@
+        assert(self.b % self.k == 0)
+
+        # compute n and store as member variable
+        self.n = self.b // self.k
```
nit and opinionated: whenever I see `//` in python, my mind gravitates the possibility that the denominator may not be divisible by the numerator (otherwise, why would we need `//` instead of `/`). because, by definition, APoT requires b % k == 0, it might mitigate overthinking on a reader's part if `/` was used here

asl3 replied:
i left it as `//` here because when i change to `/` i get a TypeError that n could be a float (but was defined as an int above). since i am asserting that b % k == 0 right above, hopefully this is fine?

dzdang reviewed (approved):
had 1 last nit but otherwise looks god to me

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Add param calcs and tests for APoT observers"


### Summary:
Previously, there was no support for calculating nonuniform observer parameters.
This PR introduces a method to calculate parameters for nonuniform observers (including gamma, quantization levels, and level indices) as well as test cases.
This implementation is based on the Additive Powers of Two (APoT) quantization method discussed in this paper: https://arxiv.org/pdf/1909.13144.pdf. The quantization parameters were calculated according to section 2.2 of the paper, and the quantization levels were calculated according to equation (5) in section 2.2.
<img width="477" alt="Screen Shot 2022-06-09 at 8 18 03 AM" src="https://user-images.githubusercontent.com/68875504/172844979-ec48ea54-e789-4470-9168-d264362beaa4.png">

### Test Plan:
python pytorch/test/quantization/core/experimental/test_nonuniform_observer.py

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
